### PR TITLE
Update winit dependency to 0.29

### PR DESCRIPTION
**Pull request by Vrixyz**
# Objective

- Update winit dependency to 0.29

## Changelog

- Changed most KeyCode values, they now represent physical keys (and could infer their scancode) rather than logical
  - [Behind other PR](https://github.com/Vrixyz/bevy/pull/3): new `KeyLogic::Logic(Key)`, with "regular" characters behind a SmolStr.
- Use `web-time` rather than `instant` crate. (https://github.com/rust-windowing/winit/pull/2836)
- winit did split `run_return` in `run_onDemand` and `pump_events`, I did the same change in bevy_winit and used `pump_events`.
- Removed `return_from_run` from `WinitSettings` as `winit::run` now returns on supported platforms.
  - I left the example "return_after_run" as  I think it's still useful. 
  - This winit change is done partly to allow to create a new window after quitting all windows: https://github.com/emilk/egui/issues/1918 ; this PR doesn't address.

## Call for help
- ❓  I'd like more feedbacks on https://github.com/Vrixyz/bevy/pull/3
- ❓ I noticed a licence file in the winit file from which we're copying most our data types: https://github.com/rust-windowing/winit/blob/master/src/keyboard.rs ; should we do something about it ?

<details><summary>Answered interrogations</summary>
<p>

#### API complexity
I'm not sure about this new API, it seems cumbersome to do https://github.com/bevyengine/bevy/pull/8745#discussion_r1216661411
```diff rs
- keycode.is_pressed(KeyCode::KeyQ)
+ key.is_pressed(Key::Character("q".into()))
```
   - ✅ Leverage Into<KeyLogic> to support `key.is_pressed("q")`

### Copy incompatible with SmolStr
We should add a logical `Key` based on winit's [Key](https://github.com/rust-windowing/winit/blob/31ebc5caf4ac43fe3bb334f66e715e5dfd95bea6/src/keyboard.rs#L667), but bevy's `KeyboardInput` is Copy, but SmolStr is not, so I'm not sure how to proceed ?
  - remove Copy ? I feel like it's a bad idea because events are usually Copy.
    - ✅ remove copy when problematic
  - use Cow or something..? 
    - ✅ use SmolStr with added reflection
  - ~add Copy upstream to SmolStr (I think that won't pass, if at all possible)~
  - ~use another type ? plain String...~
  - ~use our curated list of supported characters/string as enum, but that doesn't sound great (because chinese characters, copy/paste/IME probably wouldn't be supported this way.~
  - ✅ stick to copying winit's types as much as possible

</p>
</details> 

## Blockers
- Because this PR is big already, I separated:
  - [ ] https://github.com/Vrixyz/bevy/pull/3
- [x] [`winit`](https://github.com/rust-windowing/winit) 0.29
- [x] [`accesskit_winit`](https://github.com/AccessKit/accesskit): to update to winit 0.29 ; [ok](https://github.com/AccessKit/accesskit/releases/tag/accesskit_winit-v0.16.0)

## Follow up

I'd like to avoid bloating this PR, here are a few follow up tasks worthy of a separate PR, or new issue to track them one this PR is closed:

- [ ] NamedKey integration on bevy_input: https://github.com/rust-windowing/winit/pull/3143  introduced a new NamedKey variant. I implemented it only on the converters but we'd benefit making the same changes to bevy_input (that means also more changes to examples, etc.).
- [ ] https://github.com/bevyengine/bevy/pull/9905 added a workaround on a bug fixed by winit 0.29.
- [ ] on Android, https://github.com/rust-mobile/android-activity/issues/141 is pulling an unneeded dependency if we're keeping raw-window-handle 0.5 (we don't want to update to 0.6 [until wgpu does](https://github.com/gfx-rs/wgpu/pull/4202), I guess we don't want to follow [the maintainance burden of supporting multiple raw window handle versions](https://github.com/rust-windowing/winit/pull/3126)).
- [ ] Allow receiving multiple characters through `ReceivedCharacter`: [This PR only sends the first character](https://github.com/bevyengine/bevy/blob/c5de84734f13c8d3357d63c7f3c71030f985980c/crates/bevy_winit/src/lib.rs#L485-L490), [comment](https://github.com/bevyengine/bevy/pull/8745/files#r1387011454)
- [ ] remove `instant` dependency, [replaced by](https://github.com/rust-windowing/winit/pull/2836) `web_time`)
  - [ ] fastrand 2.0 did the switch, but following dependencies did not update to fastrand 2.0:
    - [x] [`async-executor`](https://github.com/smol-rs/async-executor)
    - [ ] [`futures-lite`](https://github.com/smol-rs/futures-lite)

## Migration Guide

This PR should have one.

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
> We should add a logical Key based on winit's [Key](https://github.com/rust-windowing/winit/blob/31ebc5caf4ac43fe3bb334f66e715e5dfd95bea6/src/keyboard.rs#L667), but bevy's KeyboardInput is Copy, but SmolStr is not, so I'm not sure how to proceed ?

IMO we should probably just remove the `Copy` bound here. Exposing this information is important, and all the other workarounds are worse than just having to call .clone().

**Comment by daxpedda** [comment]
The Web backend in Winit had quite a bit of churn lately, let me know if you uncover any bugs or issues.

**Comment by Vrixyz** [comment]
Locally, I'm exploring:
- For logical key, I make bevy_input declare a `Key<String>` populated from winit's `Key<SmolStr>`, it's easier to implement for me so I don't get FromReflect errors because of `SmolStr`.
- In `KeyboardInput`, ^ That logical key is now a field `logical_key: Key`, and replaces `key_code: ScanCode`
- The type `ScanCode` is replaced by a `KeyCode`, and rename `KeyboardInput`'s `scan_code` to `key_code`
  - ⚠️  I think it's wrong because for bevy user's perspective it means `key_code` switches from being the logical key to the physical key, and could be missed. That makes our wording more consistent with winit's though.

**Comment by Vrixyz** [comment]
> IMO we should probably just remove the `Copy` bound here. Exposing this information is important, and all the other workarounds are worse than just having to call .clone().

Unfortunately this option proves to be difficult, bevy's input system involves `app.init_resource::<Input<Key>>()`, where `Input<T:Copy>`, I'm not sure how to proceed.

**Comment by alice-i-cecile** [comment]
Removing the `Copy` bound from `Input` is totally fine: I don't think there's a strong reason to require that.

File: examples/input/keyboard_input.rs [hunk 1]
```diff
@@ -10,16 +10,16 @@ fn main() {
 }
 
 /// This system prints 'A' key state
-fn keyboard_input_system(keyboard_input: Res<Input<KeyCode>>) {
-    if keyboard_input.pressed(KeyCode::A) {
+fn keyboard_input_system(keyboard_input: Res<Input<Key>>) {
+    if keyboard_input.pressed(Key::Character("a".into())) {
```
Comment (Vrixyz) [inline examples/input/keyboard_input.rs:14]: latest commit 54483f9389fb4c21c2632e13251d23e82fec1f68 breaks other examples, as it's quite a bit of changes to do, I'd like to have an agreement on the strategy to apply there, creating a `String` each time an input is even tested (so each frame, multiple times), sounds bad for performance.

there might be benchmarks about it I could try running 🤔 

Ideally, we would want to bring SmolStr in dependencies, and provide a FromReflect implementation for that. Would that be needed for this PR or could it be done in another ?

Comment (Vrixyz) [reply examples/input/keyboard_input.rs:14]: current strategy is to implement Reflect* for SmolStr : https://github.com/bevyengine/bevy/pull/8771 ; until that SmolStr is merged, this winit PR is on hold.

**Comment by Vrixyz** [comment]
@alice-i-cecile I think this PR can be removed from 0.11 milestone, because winit doesn't plan to release 0.29 until august.

**Comment by simbleau** [comment]
You're doing god's work here 🙏

File: crates/bevy_winit/src/lib.rs [hunk 2]
```diff
@@ -561,8 +605,20 @@ pub fn winit_runner(mut app: App) {
...
+                        let new_logical_height = (window.resolution.height()
+                            / window.scale_factor() as f32)
+                            * window.resolution.height() as f32;
+                        //
+
+                        /*
+                        let new_logical_width =
+                            (inner_size_writer.width as f64 / new_factor) as f32;
+                        let new_logical_height =
+                            (inner_size_writer.height as f64 / new_factor) as f32;*/
```
Comment (Vrixyz) [inline crates/bevy_winit/src/lib.rs:621]: commenting just so it's obvious it's unfinished

File: crates/bevy_winit/src/lib.rs [hunk 3]
```diff
@@ -572,9 +628,15 @@ pub fn winit_runner(mut app: App) {
...
-                            .set_physical_resolution(new_inner_size.width, new_inner_size.height);
+                        // TODO: ThierryBerger: most likely wrong..
+                        window.resolution.set_physical_resolution(
+                            new_logical_width as u32,
+                            new_logical_height as u32,
+                        );
+
+                        /* window
+                        .resolution
+                        .set_physical_resolution(new_inner_size.width, new_inner_size.height);*/
```
Comment (Vrixyz) [inline crates/bevy_winit/src/lib.rs:639]: commenting just so it's obvious it's unfinished

File: crates/bevy_winit/src/lib.rs [hunk 4]
```diff
@@ -783,7 +845,7 @@ pub fn winit_runner(mut app: App) {
 
     trace!("starting winit event loop");
     if return_from_run {
-        run_return(&mut event_loop, event_handler);
+        pump_events(&mut event_loop, event_handler);
```
Comment (Vrixyz) [inline crates/bevy_winit/src/lib.rs:848]: I'm nooot so sure about it, see https://github.com/rust-windowing/winit/pull/2767 for detailed winit API changes.

File: crates/bevy_winit/src/system.rs [hunk 5]
```diff
@@ -298,10 +298,10 @@ pub(crate) fn changed_windows(
             }
 
             if window.ime_position != cache.window.ime_position {
-                winit_window.set_ime_position(LogicalPosition::new(
-                    window.ime_position.x,
-                    window.ime_position.y,
-                ));
+                winit_window.set_ime_cursor_area(
+                    LogicalPosition::new(window.ime_position.x, window.ime_position.y),
+                    PhysicalSize::new(10, 10),
```
Comment (Vrixyz) [inline crates/bevy_winit/src/system.rs:303]: 👀  I think I had a reason about hardcoding `(10,10)`, but that probably should benefit a comment, or a fix.

File: crates/bevy_winit/src/winit_config.rs [hunk 6]
```diff
@@ -28,7 +28,7 @@ pub struct WinitSettings {
     /// The runner will panic if this is set to `true` on other platforms.
     ///
     /// [`run`]: https://docs.rs/winit/latest/winit/event_loop/struct.EventLoop.html#method.run
-    /// [`run_return`]: https://docs.rs/winit/latest/winit/platform/run_return/trait.EventLoopExtRunReturn.html#tymethod.run_return
+    /// [`run_ondemand`]: https://docs.rs/winit/0.29.1-beta/winit/platform/run_ondemand/trait.EventLoopExtRunOnDemand.html#tymethod.run_ondemand
```
Comment (Vrixyz) [inline crates/bevy_winit/src/winit_config.rs:31]: wrong docs, I chose to go with pump_events ; not sure what's the best though, I gotta read https://github.com/rust-windowing/winit/issues/2900 more attentively.

Comment (Vrixyz) [reply crates/bevy_winit/src/winit_config.rs:31]: Fixed to `run_ondemand` ; but:

`run() -> !` has been changed to a `Result` https://github.com/rust-windowing/winit/pull/2767/files#diff-f5a20b394a5ba72e0e625bac291ac8e2fa04a59b2801dae39b277499d14729fdR315 ; I think it makes the example `return_after_run` (and its property `return_from_run`) useless, any feelings about that ?

Should I straight up remove that example and exposed `return_from_run` from `WinitSettings` ?

File: crates/bevy_input/src/keyboard.rs [hunk 7]
```diff
@@ -89,375 +124,1222 @@ pub fn keyboard_input_system(
...
+    Eject,
+    /// Sometimes labelled <kbd>My Computer</kbd> on the keyboard
+    LaunchApp1,
+    /// Sometimes labelled <kbd>Calculator</kbd> on the keyboard
+    LaunchApp2,
+    LaunchMail,
+    MediaPlayPause,
+    MediaSelect,
+    MediaStop,
+    MediaTrackNext,
```
Comment (Vrixyz) [inline crates/bevy_input/src/keyboard.rs:440]: all those are complaining about lack of docs, thougths @alice-i-cecile ? should I add a boring comment to each ? I'm not tooo confident with finding the correct wording for each so I'm worried about adding a false sense of completeness. Maybe worth doing in a follow up PR ?

Comment (alice-i-cecile) [reply crates/bevy_input/src/keyboard.rs:440]: Just a boring doc comment is fine for now :)

Comment (Vrixyz) [reply crates/bevy_input/src/keyboard.rs:440]: it's also a copypaste from https://github.com/rust-windowing/winit/blob/67b041e231ff435e5dfb7e2b8289bc96ba66069b/src/keyboard.rs#L878

(I'd like to raise concern about the licensing, there is one on top of that file ⚖️).

As it's a copy/paste, it might make the maintenance easier to either upstream some docs and/or allow no docs in there 🤔

**Comment by maniwani** [comment]
All of our examples should definitely use the physical `KeyCode` instead of the logical `Key`.[^1] We only used `Key` before because there was no equivalent to `KeyCode`. We've been waiting for `winit` to have a "physical key" enum since #862.

In general, players don't want their keyboard layout to mess up their key bindings. I personally type using Dvorak, but I have to keep a toggle to switch between Dvorak and QWERTY just to play games that (frustratingly) match on the logical output.

If a game's default key bindings for movement are WASD, it should bind the WASD keys in the standard QWERTY position regardless of whatever layout I have enabled.

[^1]: The true best practice is for a game to exclusively use "actions" (e.g. `Jump`) in its internal logic, with a library like [Steam Input][1] or [LWIM][2] being responsible for binding them to physical keys (or other device inputs). In the future, `bevy_input` should become more like those libraries.

[1]: https://partner.steamgames.com/doc/features/steam_controller/concepts
[2]: https://github.com/Leafwing-Studios/leafwing-input-manager

**Comment by Vrixyz** [comment]
There's quite a lot of changed API since winit-beta1 ; it'll take a bit of time digesting those. one interesting point is raw-window-handle being optionally updated to 0.6 ; I'm keeping it under 0.5 until we use a wgpu version which supports it(https://github.com/gfx-rs/wgpu/pull/4202).

**Comment by alice-i-cecile** [comment]
No longer blocked following the winit 0.29 release 🎉

File: crates/bevy_winit/src/lib.rs (hunk 4 above)
Comment (rib) [reply crates/bevy_winit/src/lib.rs]: mmmm, no, I doubt you want to use `pump_events()` in bevy.

You should be able to either use the regular `run` API now (which supports returning now - except on iOS/Web) or else use `run_on_demand()` which behave a lot more like `run_return` did.

`pump_events` is for cases where your game engine implements it's own external loop which would also need to have some mechanism for throttling that loop.

**Comment by simbleau** [comment]
🙏

**Comment by Vrixyz** [comment]
> 🙏
> 
> @Vrixyz for those who may stuble upon this, how is this issue/pr related to fixing the bevy_egui issue where non-qwerty keyboards can't type symbols or special characters on wasm32?

To answer "in short": I tracked it to the example char_input_events on bevy wasm, and noticed such special characters were not printed. winit refactored the keyboard input and I believe that's a good step towards fixing that issue.

For a longer context: https://github.com/mvlabat/bevy_egui/pull/178#issuecomment-1571038329

**Comment by devil-ira** [comment]
[`futures-lite`](https://github.com/smol-rs/futures-lite) has also updated to `fastrand` 2.0 since

File: crates/bevy_input/src/keyboard.rs [hunk 8]
```diff
@@ -89,375 +124,1264 @@ pub fn keyboard_input_system(
...
+    /// A macOS "scancode". There does not appear to be any direct analogue to either keysyms or
+    /// "virtual-key" codes in macOS, so we report the scancode instead.
+    MacOS(u16),
+    /// A Windows "virtual-key code".
+    Windows(u16),
+    /// An XKB "keysym".
+    Xkb(u32),
+    /// A "key value string".
+    // TOCLEAN: Thierry: this is originally SmolStr, but difficult to translate to FromReflect.
+    Web(SmolStr),
```
Comment (devil-ira) [inline crates/bevy_input/src/keyboard.rs:622]: Is this comment up-to-date?

File: crates/bevy_window/src/cursor.rs [hunk 9]
```diff
@@ -20,78 +20,130 @@ use bevy_reflect::{ReflectDeserialize, ReflectSerialize};
...
-    /// An arrow. This is the default cursor on most systems.    
-    Arrow,
-    /// Indicates something is to be moved.    
-    Move,
-    /// Indicates text that may be selected or edited.    
-    Text,
-    /// Program busy indicator.    
-    Wait,
-    /// Help indicator (often rendered as a "?")    
+
```
Comment (devil-ira) [inline crates/bevy_window/src/cursor.rs:26]: Not sure about all these empty lines.

**Review by devil-ira — COMMENTED** [review]
Reviewed! This is a great step forward for input handling c:

File: crates/bevy_winit/src/system.rs [hunk 10]
```diff
@@ -174,7 +174,7 @@ pub(crate) fn changed_windows(
                     window.resolution.physical_width(),
                     window.resolution.physical_height(),
                 );
-                winit_window.set_inner_size(physical_size);
+                let _ = winit_window.request_inner_size(physical_size);
```
[the remark this answers is not in the document]
Comment (Vrixyz) [reply crates/bevy_winit/src/system.rs:177]: [(link to docs)](https://docs.rs/winit/0.29.3/winit/window/struct.Window.html#method.request_inner_size)
" **may** not be generated " 🙁 ; I don't like uncertainty...

this seems to [only affect wayland?](https://github.com/rust-windowing/winit/blob/21701a33de20775fd1fb11ca75b23bf5438a7acf/src/platform_impl/linux/wayland/window/mod.rs#L315-L320) ; but the exact path to the **maybe not** generated event is quite hard to follow 🤔...

I'll follow the trail to winit and might open an issue because I'd like to handle differently:
- Some(newValue): generate an event resize [like we do when reacting from events](https://github.com/bevyengine/bevy/blob/66541a085608b60da4c24bdda1861c9904fad375/crates/bevy_winit/src/lib.rs#L477C38-L486) ⚠️ the "**may**" from the docs may cause the event to be sent twice.
- Some(oldValue): do not generate an event resize ⚠️ I cannot differentiate it from a new value, so I'd send "useless" resize. Well I guess we're not supposed to send a resize event in those platforms anyway but still.
- None: event will be sent afterward, or not ✅ 

maybe the docs is wrong and should read "**will** not be generated", so I can handle it with certainty 🤞.

File: examples/app/return_after_run.rs [hunk 11]
```diff
@@ -1,25 +1,16 @@
...
-//! This example demonstrates the use of `WinitSettings::return_from_run` to
-//! require that `App::run()` *does* return but this is not recommended. Be sure
-//! to read the documentation on both `App::run()` and `WinitSettings::return_from_run`
-//! for caveats and further details:
-//!
-//! - <https://docs.rs/bevy/latest/bevy/app/struct.App.html#method.run>
-//! - <https://docs.rs/bevy/latest/bevy/winit/struct.WinitSettings.html#structfield.return_from_run>
+//! not recommended because:
+//! - `App::run()` will never return on iOS and Web.
+//! - It is not possible to recreate a window afer the event loop has been terminated.
```
[the remark this answers is not in the document]
Comment (Vrixyz) [reply examples/app/return_after_run.rs:6]: I believe so: related links:
- https://github.com/rust-windowing/winit/issues/2431
- https://github.com/rust-windowing/winit/pull/2767

File: examples/audio/audio_control.rs [hunk 12]
```diff
@@ -39,9 +39,9 @@ fn pause(keyboard_input: Res<Input<KeyCode>>, music_controller: Query<&AudioSink
 
 fn volume(keyboard_input: Res<Input<KeyCode>>, music_controller: Query<&AudioSink, With<MyMusic>>) {
     if let Ok(sink) = music_controller.get_single() {
-        if keyboard_input.just_pressed(KeyCode::Plus) {
+        if keyboard_input.just_pressed(KeyCode::NumpadAdd) {
```
[the remark this answers is not in the document]
Comment (Vrixyz) [reply examples/audio/audio_control.rs:42]: Not sure; I'll use equal.

File: examples/audio/audio_control.rs [hunk 13]
```diff
@@ -39,9 +39,9 @@ fn pause(keyboard_input: Res<Input<KeyCode>>, music_controller: Query<&AudioSink
...
             sink.set_volume(sink.volume() + 0.1);
-        } else if keyboard_input.just_pressed(KeyCode::Minus) {
+        } else if keyboard_input.just_pressed(KeyCode::NumpadSubtract) {
```
Comment (Vrixyz) [inline examples/audio/audio_control.rs:44]: same as https://github.com/bevyengine/bevy/pull/8745/files/66541a085608b60da4c24bdda1861c9904fad375#r1387069117 ; but <kbd>Minus</kbd> still exists so I'll use <kbd>Minus</kbd>.

File: crates/bevy_winit/src/system.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (Vrixyz) [reply crates/bevy_winit/src/system.rs:177]: Thanks for bringing up resizing, I think that PR has a small regression, as window is not redrawn during resizing.

<details><summary>My PR</summary>
<p>

https://github.com/bevyengine/bevy/assets/2290685/0971969a-cbba-4ad4-a40a-654f322bf3b4

</p>
</details> 

<details><summary>On main</summary>
<p>

(rev 0cc11791b9a55f5d310ee754c9f7254081e7138a)

https://github.com/bevyengine/bevy/assets/2290685/e7cc36b5-c0c8-4797-bc1b-a8e4bbcab938

</p>
</details>

File: crates/bevy_window/src/cursor.rs (hunk 9 above)
Comment (devil-ira) [reply crates/bevy_window/src/cursor.rs:26]: Copying the message in winit's 'keyboard.rs' verbatim to our 'keyboard.rs' file(Where `Key` and `KeyCode` are defined) will satisfy the W3C license :)

File: crates/bevy_winit/src/lib.rs [hunk 14]
```diff
@@ -529,10 +492,19 @@ pub fn winit_runner(mut app: App) {
...
-                    WindowEvent::KeyboardInput { ref input, .. } => {
-                        event_writers
-                            .keyboard_input
-                            .send(converters::convert_keyboard_input(input, window_entity));
+                    WindowEvent::KeyboardInput { ref event, .. } => {
+                        let keyboard_event =
+                            converters::convert_keyboard_input(event, window_entity);
+                        if let bevy_input::keyboard::Key::Character(c) = &keyboard_event.logical_key
+                        {
+                            if let Some(first_char) = c.chars().next() {
```
[the remark this answers is not in the document]
Comment (Vrixyz) [reply crates/bevy_winit/src/lib.rs:485]: I.. guess not, but sending the whole SmolStr within ReceivedCharacter seemed a big change when I was on it.

I can see multiple solutions:
- sending SmolStr.
- sending several ReceivedCharacter(char)
- splitting ReceivedCharacter and ReceivedMultipleCharacters ?

I thought this implementation was the closest to current main behaviour, so that might be OK as a first approach.
I'm writing that in "Follow Up" for the time being.

File: examples/audio/audio_control.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (alice-i-cecile) [reply examples/audio/audio_control.rs]: These were swapped in `winit` by mistake IIRC :(

File: crates/bevy_window/src/cursor.rs (hunk 9 above)
Comment (Vrixyz) [reply crates/bevy_window/src/cursor.rs:26]: from https://www.w3.org/copyright/software-license-2002/ ; I think we'd need to also include the changes made. they are pretty straightforward: 
- add Reflect capabilities
- use another feature name for serialization
- maybe not using non_exhaustive for enums?

But I lack knowledge on how to write that. I'll copy the licenses as I believe it's better than nothing :)

**Comment by Vrixyz** [comment]
I'm closing this PR in favor of https://github.com/bevyengine/bevy/pull/10702 ; which builds on this one but it's time to get a shorter comment thread

File: crates/bevy_winit/src/winit_windows.rs [hunk 15]
```diff
@@ -194,7 +194,7 @@ impl WinitWindows {
             use winit::platform::web::WindowExtWebSys;
 
             if window.canvas.is_none() {
-                let canvas = winit_window.canvas();
+                let canvas = winit_window.canvas().unwrap();
```
[the remark this answers is not in the document]
Comment (Vrixyz) [reply crates/bevy_winit/src/winit_windows.rs:197]: I think it's alright to call unwrap on that, I believe it's instantiated at [the same time of a window](https://github.com/rust-windowing/winit/blob/master/src/platform_impl/web/window.rs#L44), so it might fail if you're doing convoluted things like modifying the canvas from javascript.

But that's not a rabbit hole I'm ready to follow yet. I'll make a note about it.

[the remark this answers is not in the document]
Comment (daxpedda) [reply crates/bevy_winit/src/winit_windows.rs:197]: > Only returns the canvas if called from inside the window.

Apparently this didn't turn out very clear. This will only return `None` if called from a Web worker, it will always succeed if called from the "main thread" (the "window" context).
